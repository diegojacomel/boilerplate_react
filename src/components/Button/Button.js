// Modules
import React from "react";
import PropTypes from 'prop-types';

// Components
import Icon from '../Icon/Icon';
import { Link } from 'react-router-dom';

// Styles
import styled from "styled-components";

const OutlineButton = (theme, outline, color) => {
    if (outline) {
        return `
            border: 1px solid ${color};
            border-bottom: ${({ showShadow }) => showShadow ? `none` : ``};
            background-color: ${theme.colors.white};
            color: ${color};
            &:hover {
                background-color: ${color};
                color: ${theme.colors.white};
            }
    `;
    } else {
        return `
            background-color: ${color};
            color: ${theme.colors.white};
    `;
    }
};

const SizeButton = (props) => {
    switch (props.size) {
        case 'xg':
            return `
                padding: ${props.theme.spacing.lg} ${props.theme.spacing.xg};
            `
        case 'lg':
            return `
                padding: ${props.theme.spacing.md} ${props.theme.spacing.xlg};
            `
        case 'md':
            return `
                padding: ${props.theme.spacing.sm} ${props.theme.spacing.lg};
            `
        case 'sm':
            return `
                padding: ${props.theme.spacing.xsm} ${props.theme.spacing.md};
            `
        case 'xs':
            return `
                padding: ${props.theme.spacing.xs} ${props.theme.spacing.sm};
            `
        default:
            return `
                padding: ${props.theme.spacing.sm} ${props.theme.spacing.lg};'
        `
    }
}

const ButtonStyle = styled('button')`
    text-align: center;
    outline: none;
    border: none;
    line-height: 1;
    transition: all 0.3s;
    margin-right: 15px;
    white-space: nowrap;
    width: ${({ block }) => block ? '100%' : 'initial'};
    border-radius: ${({theme, rounded}) => theme.rounded[rounded]};
    font-size: ${({theme, size}) => theme.fontSize[size]};
    ${({ theme, outline, color }) => OutlineButton(theme, outline, theme.colors[color])};
    ${props => SizeButton(props)};
    position: relative;
    top: 0;
    transition: all 0.2s;
    box-shadow: ${({ theme, color, showShadow }) => showShadow ? `0 5px 0 ${theme.colors.active[color]}` : `0 0 0 ${theme.colors.active[color]}`};
    cursor: pointer;
    &:hover {
        ${({ theme, outline, color }) => OutlineButton(theme, outline, theme.colors.hover[color])};
        
    }
    :active {
        top: ${({ showShadow }) => showShadow ? `5px` : '0'};
        box-shadow: 0 0 0 ${({ theme, color }) => theme.colors.active[color]};
    }
    &:last-child {
        margin-right: 0;
    }
    & > span {
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-direction: ${({ iconLeft }) => iconLeft ? 'row-reverse' : 'row'};
        & > svg {
            margin-left: ${({ iconLeft }) => !iconLeft ? '10px' : '0'};
            margin-right: ${({ iconLeft }) => iconLeft ? '10px' : '0'};
        }
    }
`;

const LinkStyle = styled(Link)`
    text-decoration: none;
    color: inherit;
`

const Button = ({ children, type, color, outline, size, rounded, block, icon, iconSize, iconColor, iconLeft, disabled, onClick, linkTo, showShadow }) => (
    <ButtonStyle
        type={type}
        color={color}
        outline={outline}
        size={size}
        rounded={rounded}
        block={block}
        iconLeft={iconLeft}
        onClick={onClick}
        disabled={disabled}
        showShadow={showShadow}
        id="testButton"
    >
        {!!linkTo
            ?
            <LinkStyle id="testLinkButton" to={linkTo}>
                <span>
                    {children}
                    {icon
                        ?
                        <Icon tag={icon} iconSize={iconSize} iconColor={iconColor} />
                        :
                        null
                    }
                </span>
            </LinkStyle>
            :
            <span>
                {children}
                {icon
                    ?
                    <Icon tag={icon} iconSize={iconSize} iconColor={iconColor} />
                    :
                    null
                }
            </span>
        }
    </ButtonStyle>
);

Button.propTypes = {
    type: PropTypes.string,
    color: PropTypes.string,
    outline: PropTypes.bool,
    size: PropTypes.string,
    rounded: PropTypes.string,
    block: PropTypes.bool,
    icon: PropTypes.string,
    iconSize: PropTypes.string,
    iconColor: PropTypes.string,
    disabled: PropTypes.bool,
    linkTo: PropTypes.string,
    onClick: PropTypes.func,
    showShadow: PropTypes.bool
}

Button.defaultProps = {
    type: 'button',
    color: 'default',
    outline: false,
    size: 'md',
    rounded: 'lg',
    block: false,
    disabled: false,
    linkTo: null,
    onClick: () => console.log('clicked button'),
    showShadow: false
}

export default Button;