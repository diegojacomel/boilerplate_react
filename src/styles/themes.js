// Scale
import { pxToRem } from '../utils/scale';

// Themes
const primaryTheme = {
    colors: {
        primary: "#1a81ff",
        danger: "#ff3333",
        info: "#05cdfa",
        success: "#159d79",
        default: "#373f49",
        white: "#ffffff",
        black: "#1a1a1a",
        greyLight: "#f2f2f2",
        greyMedium: "#999999",
        greyDark: "#333333",
        hover: {
            primary: "#54a0ff",
            danger: "#ff6b6b",
            info: "#48dbfb",
            success: "#18bb90",
            default: "#576574",
            white: "#ffffff",
            black: "#000000"
        },
        active: {
            primary: "#0069ed",
            danger: "#9e0000",
            info: "#05b2d8",
            success: "#11775c",
            default: "#111111",
            white: "#f2f2f2",
            black: "#1a1a1a"
        }
    },
    spacing: {
        none: "0",
        xxs: pxToRem(3),
        xs: pxToRem(5),
        xsm: pxToRem(8),
        sm: pxToRem(10),
        md: pxToRem(15),
        lg: pxToRem(20),
        xlg: pxToRem(25),
        xg: pxToRem(30),
        xxg: pxToRem(50)
    },
    fontSize: {
        none: "0",
        xs: pxToRem(12),
        sm: pxToRem(14),
        md: pxToRem(16),
        lg: pxToRem(18),
        xg: pxToRem(20),
    },
    rounded: {
        none: "0",
        sm: "3px",
        mdLeft: "5px 0 0 5px",
        mdRight: "0 5px 5px 0",
        md: "5px",
        lg: "10px",
        full: "100%"
    }
}

export {
    primaryTheme
}