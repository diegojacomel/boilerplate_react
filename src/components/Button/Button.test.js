// Modules
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

// Theme
import { primaryTheme } from '../../styles/themes'; 

// Component
import Button from './Button';

describe('Button', () => {
    const wrapper = shallow(<Button>Click here</Button>);

    test('It exist', () => {
        expect(wrapper.find('#testButton')).toExist();
    })

    test('At least one', () => {
        expect(wrapper).toContainMatchingElement('#testButton');
    })

    test('Have display name', () => {
        expect(wrapper.find('#testButton')).toHaveDisplayName('styled.button');
    })

    test('Match selector', () => {
        expect(wrapper.find('#testButton')).toMatchSelector('#testButton');
    })

    test('Have props', () => {
        expect(wrapper.find('#testButton')).toHaveProp('type');
        expect(wrapper.find('#testButton')).toHaveProp('color');
        expect(wrapper.find('#testButton')).toHaveProp('outline');
        expect(wrapper.find('#testButton')).toHaveProp('size');
        expect(wrapper.find('#testButton')).toHaveProp('rounded');
        expect(wrapper.find('#testButton')).toHaveProp('block');
        expect(wrapper.find('#testButton')).toHaveProp('disabled');
        expect(wrapper.find('#testButton')).toHaveProp('onClick');
    })

    it('Render correctly', () => {
        const tree = renderer.create(
            <ThemeProvider theme={primaryTheme}>
                <Button>
                    Click here
                </Button>
            </ThemeProvider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    })
})