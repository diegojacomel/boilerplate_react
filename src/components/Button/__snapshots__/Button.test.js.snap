// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render correctly 1`] = `
<button
  className="sc-bwzfXH hqHnCz"
  color="default"
  disabled={false}
  id="testButton"
  onClick={[Function]}
  size="md"
  type="button"
>
  <span>
    Click here
  </span>
</button>
`;
